### ISSUE TITLE
CSV Importer - fwrite(): write of xx bytes failed with errno=9 Bad file descriptor

### ISSUE THREAD

**Issue by Moc** [firefly-iii/firefly-iii#3646]
I am running Firefly III version 5.3.3

**Description**
After updating to the most recent CSV Importer version, I get this error. Tried a clean install same error. Tried doing research but haven't found anything other than perhaps this PHP "bug" (https://bugs.php.net/bug.php?id=78482) but I am not sure if it is applicable to me.  I figure it is related to write permissions but I don't know which folder it needs to write to then...

I tried playing around with line 103 of `C:\webdev\www\csv-importer-new\vendor\monolog\monolog\src\Monolog\Handler\StreamHandler.php`, but no luck. 
```php
            $this->stream = fopen($this->url, 'a');
```
(changing the `a` to `a+` for instance) . 

Not sure what else I can do. Not even sure if this is even an issue in CSV importer or in the Monolog package. It may well be something that I have to fix on my environment but I would appreciate some pointers.. 

```
[2020-08-03 17:31:43] local.DEBUG: Now at App\Http\Controllers\IndexController::index  
[2020-08-03 17:31:43] local.ERROR: fwrite(): write of 88 bytes failed with errno=9 Bad file descriptor {"exception":"[object] (ErrorException(code: 0): fwrite(): write of 88 bytes failed with errno=9 Bad file descriptor at C:\\webdev\\www\\csv-importer-new\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\StreamHandler.php:134)
```

**Extra info**
Windows (localhost)  
PHP 7.4.8 
With Write access to C:\Windows\Temp

**Comment by JC5** [comment]
I hope somebody else knows because I know nothing about Windows.

**Comment by sephrat** [comment]
I'm no real expert but I'll try to help.
I can't reproduce on my Windows setup. Does it fail on home page or on a later step?
In the former case, it could be the log file. I've noticed the path is a bit off "C:\xampp\htdocs\csv-importer-git\storage\logs/csv-import.log" (slash instead of backslash) but it's handled fine by PHP. And I'm assuming there's no reason for there to be a permission issue.
Why did you mention C:\Windows\Temp? By default you need admin rights to access the directory. I was still able to import a CSV a file from that directory, but your error message is about writing to a file so that could be it.

**Comment by Moc** [comment]
@sephrat I appreciate your help, thank you! 

It fails on the homepage. A log file is present there (and re-created when deleted) which contains the debug information with the error I posted. 

I mentioned the ` C:\Windows\Temp` folder because this was the root of another (permission) error: https://github.com/firefly-iii/firefly-iii/issues/2676

I am still trying to deduce what has changed, because the CSV Importer was working before on this environment. Only after updating to a more recent version, the error appeared. I can't remember which version I was on before updating, but I will try previous versions asap.

**Comment by Moc** [comment]
Just tested with CSV Importer 1.0.15, which is working fine with PHP 7.3. If I use CSV importer 1.0.15 with PHP 7.4, the same issue occurs. This basically tells me it is an issue with the PHP 7.4 on Windows (e..g https://bugs.php.net/bug.php?id=78482) but no idea where/how to fix this.

**Comment by Moc** [comment]
So I have an update for this:

- I recently updated Firefly itself (not CSV importer) and the same error appeared. 
- I found https://bugs.php.net/bug.php?id=79166 and https://bugs.php.net/bug.php?id=79905
- I figured I was using the *non thread safe* (NTS) version of Windows. When switching to *thread safe*, the error disappeared. 

It seems like a bug in PHP 7.4 on Windows. 
Hope this may help someone in the future.
